doc/rapiddisk.1
.TH RAPIDDISK 1 "Oct 16 2010" "Linux" "GENERAL COMMANDS"
.SH NAME
rapiddisk \- An administration tool to manage the RapidDisk RAM disk devices and RapidDisk-Cache mappings.
.SH SYNOPSIS
rapiddisk [ -h | -v ] function [ parameters ]
.SH DESCRIPTION
rapiddisk is a RapidDisk module management tool to manage RapidDisk RAM disk devices. Dynamically create, remove, resize RAM volumes and if desired, map or unmap them as a cache volume to any block device. Access those drives locally or export those volumes across an NVMe Target network.
.SS Options
.TP
-h
Invoke the help menu.
.TP
-v
Display the version number.
.SS Functions
.TP
-a
Attach RAM disk device (size in MBytes).
.TP
-b
Backend block device absolute path (for cache mapping).
.TP
-c
Input capacity for size or resize of RAM disk device (in MBytes).
.TP
-d
Detach RAM disk device.
.TP
-e
Export a RapidDisk block device as an NVMe Target.
.TP
-f
Erase all data to a specified RapidDisk device (dangerous).
.TP
-g
Do not print header.
.TP
-H
The host to export / unexport the NVMe Target to / from.
.TP
-i
Define the network interface to enable for NVMe Target exporting. Port must not already exist and interface must not be already enabled.
.TP
-j
Enable JSON formatted output.
.TP
-L
Lock a RapidDisk block device (set to read-only).
.TP
-l
List all attached RAM disk devices.
.TP
-m
Map an RapidDisk device as a caching node to another block device.
.TP
-N
List only enabled NVMe Target ports.
.TP
-n
List RapidDisk enabled NVMe Target exports.
.TP
-P
The port to export / unexport the NVMe Target to / from.
.TP
-p
Define cache policy: write-through (wt), write-around (wa) or writeback (wb) (dangerous) (default: write-through).
Writeback caching is supplied by the dm-writecache kernel module and is not intended for production use as it may result in data loss on hardware/power failure.
.TP
-q
List all system memory and block device resources.
.TP
-R
Revalidate size of NVMe export using existing RapidDisk device.
.TP
-r
Dynamically grow the size of an existing RapidDisk device.
.TP
-s
Obtain RapidDisk-Cache Mappings statistics.
.TP
-t
Define the NVMe Target port's transfer protocol (i.e. tcp, rdma or loop).
.TP
-U
Unlock a RapidDisk block device (set to read-write).
.TP
-u
Unmap a RapidDisk device from another block device.
.TP
-X
Remove the NVMe Target port (must be unused).
.TP
-x
Unexport a RapidDisk block device from an NVMe Target. To remove export to host or port, only define the host and / or port. Not defining a host or port will result in the block device being removed from the NVMe Target subsystem.
.SS Parameters (if applicable)
.TP
[size]
Specify desired size of attaching RAM disk device in MBytes.
.TP
[mode]
Write Through (wt) or Write Around (wa) for cache.
.SH EXAMPLE USAGE
.TP
rapiddisk -l
.TP
rapiddisk -l -j
.TP
rapiddisk -a 64
.TP
rapiddisk -d rd2
.TP
rapiddisk -r rd2 -c 128
.TP
rapiddisk -m rd1 -b /dev/sdb
.TP
rapiddisk -m rd1 -b /dev/sdb -p wt
.TP
rapiddisk -m rd3 -b /dev/mapper/rc-wa_sdb -p wb
.TP
rapiddisk -u rc-wt_sdb
.TP
rapiddisk -s rc-wt_sdb
.TP
rapiddisk -f rd2
.TP
rapiddisk -L rd2
.TP
rapiddisk -U rd3
.TP
rapiddisk -i eth0 -P 1 -t tcp
.TP
rapiddisk -i NULL -P 1 -t loop
.TP
rapiddisk -X -P 1
.TP
rapiddisk -e -b rd3 -P 1 -H nqn.host1
.TP
rapiddisk -R -b rd0
.TP
rapiddisk -x -b rd3 -P 1 -H nqn.host1
.SH MANAGING RAPIDDISK AS AN NVME TARGET 
There are a few things that need to be known when using the NVMe Target features of the RapidDisk suite.
.TP
1.
In order to map any RapidDisk device and export it in the NVMe Target framework, the nvmet and the nvmet-tcp or nvmet-rdma kernel modules must be inserted.
.TP
2.
At least one Ethernet interface will need to be configured as a target port to export the RapidDisk volume from.
.RS
.P
ex) rapiddisk -i eth -P 1 -t tcp
.RE
.TP
3.
When exporting a volume, a RapidDisk volume and a target port must be defined. If a host NQN is not defined, the administration utility will provide access to any host NQN. Note - a target can be exported across more than one target port.
.RS
.P
ex) rapiddisk -e -b rd3 -P 1
.P
If a host NQN is defined, access is restricted to only those host NQNs. Note - the following command example can be repeated multiple times to add additional host NQNs for the specified target export.
.P
ex) rapiddisk -e -b rd3 -P 1 -H nqn.host1
.RE
.TP
4.
Unexporting RapidDisk volumes looks a bit different than exporting. If a host NQN is defined for a specified target, only that NQN will be removed from accessing the exported target.
.RS
.P
ex) rapiddisk -x -b rd3 -H nqn.host1
.P
Removing all allowed host NQNs will revert access to any and all host NQNs requesting access to the target.
.P
If a target port is defined, the exported target will not be exported from the interface if one condition is met: the target has no defined allowed host NQNs.
.P
ex) rapiddisk -x -b rd3 -P 1 -H nqn.host1
.P
ex) rapiddisk -x -b rd3 -P 1
.P
And if there are no defined allowed host NQNs and the target is not being exported across any target ports, the entire target is removed from the subsystem.
.RE
.SH EXIT STATUS
rapiddisk returns a zero exit status if no error occurs during operation. A non-zero value is returned on error.
.SH AUTHORS
Original version: Petros Koutoupis (petros@petroskoutoupis.com)
.SH SEE ALSO
View the RapidDisk project page: http://www.rapiddisk.org


CHANGELOG
*** Release 4.4 ***

- kernel: Update to 4.8 and 4.9 kernels.
- build: Cleaned up Makefiles (thanks Marcel!)

*** Release 4.3 ***

- kernel: Add support for the 4.7 kernel (patch supplied by Marcel Huber)
- packaging: Updated DEB control for PHP changes between Ubuntu 14.04/16.04

*** Release 4.2-3 ***

- utility: Fixed bug in make install with nocrypt enabled.
- documentation: clean up.

*** Release 4.2-2 ***

- utility: Added more complex default DES key with backwards compatibility to legacy key.
- ha: added write around support to HA resource agents.
- documentation: corrections / clean up.

*** Release 4.2 ***

- kernel: Added Write-Around support to rapiddisk-cache.
- kernel: Fixed LINUX_VERSION check for rapiddisk-cache to accommodate changes in 3.8.3.
- utility: Added a NOCRYPT build flag.
- utility: Added user definable keys for encryption setup.

*** Release 4.1-2 ***

- kernel: Readjusted misaligned discard request check to build on kernels older than 4.3.

*** Release 4.1 ***

- kernel: Refuse misaligned discard requests.
- kernel: Convert ENOMEM to ENOSPC when cannot alloc pages.
- kernel: Added 4k physical block size attribute.

*** Release 4.0-4 ***

- Code cleanup.
- packaging: Also need to remove CONFIG_BLK_DEV_RAM_COUNT from distro specific packages.

*** Release 4.0-3 ***

- utility: Fixed all references of RapidCache to RapidDisk-Cache.

*** Release 4.0-2 ***

- utility: Updated fatfree-framework.
- utility: Addressed bug in RapidDisk REST implementation.

*** Release 4.0 ***

- Fixed libcryptsetup build error for RHEL6.
- Renamed rxdsk/rxcache modules.
- Did massive cleanup of administration utility code.
- Converted most (if not all) return codes to POSIX.1 error numbers.
- Code cleanup in RESTful API.
- RapidDisk volumes now show up as non-rotational.
- Cleaned up ioctls in both module and utilies.

*** Release 3.7 ***

- Cleaned up kernel module code.
- Fixed stack overflow bug in administration utility.
- Fixed error print statement in administration utility.
- Updated copyright years.
- Cleaned up build environment, including Makefiles.
- Fixed bug in configuring encryption on device.

*** Release 3.6-3 ***

- Updated/corrected documentation in manual page and in source.

*** Release 3.6-2 ***

- Fixed a memory leak in administration utility.
- Placed better checks before deallocating memory in administration utility.

*** Release 3.6 ***

- Updated RapidDisk modules for Linux kernel version 4.4.

*** Release 3.5-2 ***

- Updated kernel documentation.

*** Release 3.5 ***

- Forcing rxdsk driver to do the drive enumeration. Removed functionality from administration utility.
- Appropriately initializing major number variable to 0 before registering block device module.
- Administration utility now checks for sysfs entry of rxdsk and not module name in /proc/modules.
- Converted sector size input in rxdsk module to KB input. Modified administration utility to this.
- Fixed bug in check for total rxdsk devices in module.

*** Release 3.4-2 ***

- Updated Makefile for tools install to add HA scripts.
- Fixed bugs with HA scripts (both rgmanager and pacemaker).

*** Release 3.4 ***

- Added ability to autoload RapidDisk volumes during module insertion.
- Fixed bug in RapidDisk (volatile) volume size definition across 32 to 64 bit types.
- Making use of BIT() macro in the driver.
- Removed RapidDisk-NV support. It was redundant with the recently kernel integrated pmem code.

*** Release 3.3 ***

- Updated code for the 4.3 kernel.
- Cleaned up the main Makefile.
- Cleaned up entire driver code. Adjusted formatting.

*** Release 3.2 ***

- Replaced procfs management to sysfs.
- Identified & corrected a couple of memory leaks.
- Massive code cleanup (intended for kernel submission).
- Minor code optimizations (slight performance improvements).

*** Release 3.1-2 ***

- Fixed on-line menu of administration binary.
- Updated spec file to autoload modules after install.
- Corrected package description in spec file.

*** Release 3.1 ***

- Fixed memory leak and an exit on failure before removing mutex during a procfs read.
- Added RESTful test file to test API from CLI.
- Integrated encryption support via dm-crypt.
- Enabled RPM builds for Red Hat / CentOS 6 & 7.
- Enabled RapidDisk YUM repo for Red Hat / CentOS 6 & 7 support.
- Added Pacemaker and rgmanager resource files to enable HA support.

*** Release 3.0 ***

- Added NVDIMM support.
- Added RESTful API support.
- Updated administration binary and cleaned up a lot of its code.
- Removed pyRxAdm graphical wrapper.
- Fixed bug when erroring during RapidCache module insertion.
- Cleaned up RapidCache module code (removed procfs entry).

*** Release 2.13 ***

- Bug fix with rxadm binary and mapping RapidCache to pre-existing partitions.
- Added more information to RapidDisk proc_fs file.
- Addressed compilation warnings for GCC 5.1

*** Release 2.12 ***

- Updated modules for kernel 3.14

*** Release 2.11 ***

- Updated rxcache for Red Hat 6.4 (device mapper conflict)
- Addressed incorrect description of maximum number of rxdsks supported.

*** Release 2.10 ***

- Updated modules for kernel 3.10.
- Updated Makefile for cross compiling install.

*** Release 2.9.2 ***

- Addressed a bug in rxdsk print statement (wrong type). Thanks go to Neo for discovering and patching it.

*** Release 2.9.1 ***

- Minor update adding DKMS support.
- Adding support to build and install/uninstall tools separately (i.e. without modules, as in when installing with DKMS).

*** Release 2.9 ***

- Added better implementation of BLKFLSBUF ioctl to rxdsk module. This will "flush data" and truncate pages.
- Added flush command to rxadm utility.
- Added support for Linux kernel 3.9. Tested on 3.9.2.

*** Release 2.8 ***

- Cleaned up code and removed unused and unimplemented caching feature (write-around).
- Added support for Linux kernel 3.8. Tested on 3.8-rc7.

*** Release 2.7 ***

- Made some modifications to the modules' makefile.

www/rapiddisk.php
<?php

class RapidDisk {
	public $RDISK_CMD;
	
	function __construct() {
		 $this->RDISK_CMD = 'sudo -u root /sbin/rapiddisk';
	}

	/* @GET: List all RapidDisk and RapidDisk-Cache volumes with sizes (in MBs) and mappings. */
	public function listAllRapidDiskVolumes() {
                $command = implode(' ', array($this->RDISK_CMD, '--short-list'));
                exec($command, $output, $retval);
		$list = $tmp = [];
		foreach ($output as &$volume) {
			$var = explode(':', $volume);
                        if (strstr($var[0], 'Unable to locate any') != False) {
                                unset($tmp);
                                break;
			} elseif (strstr($var[0], 'rd') != False) {
				$tmp = array('rapidDisk' => $var[0], 'size' => intval($var[1]));
				array_push($list, $tmp);
			} elseif (strstr($var[0], 'rc') != False) {
				$tmp = array('rapidCache' => $var[0], 'cache' => explode(',', $var[1])[0], 'source' => explode(',', $var[1])[1]);
				array_push($list, $tmp);
			}
			unset($tmp);
		}
                $rtn = array('volumes' => $list);
		http_response_code(200); /* OK */
		print json_encode($rtn);
	}

        /* @GET: Show cache statistics of an existing RapidDisk-Cache Volume. */
        public function showRapidDiskCacheStatistics($cache) {
                $command = implode(' ', array($this->RDISK_CMD, '--stat-cache', $cache, '|grep "("|tr -s \'[[:space:]]\' \' \'|sed -e \'s/,//\'g -e \'s/^ //\' -e \'s/$/\n/\''));
                exec($command, $output, $retval);
                $rtn = array('errorCode' => $retval, 'message' => implode(' ', $output));
                http_response_code(200); /* OK */
                print json_encode($rtn);
        }

	/* @POST: Create a (volatile) RapidDisk Volume (size in MBs). */
	public function createRapidDisk($JSONRequest) {
                $nameObject = json_decode($JSONRequest, true);
		$size = $nameObject['size'];
                $command = implode(' ', array($this->RDISK_CMD, '--attach', $size));
                exec($command, $output, $retval);
                $rtn = array('errorCode' => $retval, 'message' => implode(' ', $output));
		http_response_code(200); /* OK */
		print json_encode($rtn);
	}

        /* @PUT: Resize an existing (volatile) RapidDisk Volume (size in MBs). */
        public function resizeRapidDisk($JSONRequest) {
                $nameObject = json_decode($JSONRequest, true);
                $name = $nameObject['rapidDisk'];
                $size = $nameObject['size'];
                $command = implode(' ', array($this->RDISK_CMD, '--resize', $name, $size));
                exec($command, $output, $retval);
                $rtn = array('errorCode' => $retval, 'message' => implode(' ', $output));
                http_response_code(200); /* OK */
                print json_encode($rtn);
        }

	/* @DELETE: Delete an existing RapidDisk Volume. */
	public function removeRapidDisk($disk) {
                $command = implode(' ', array($this->RDISK_CMD, '--detach', $disk));
                exec($command, $output, $retval);
                $rtn = array('errorCode' => $retval, 'message' => implode(' ', $output));
                http_response_code(200); /* OK */
                print json_encode($rtn);
	}

        /* @POST: Create a RapidDisk-Cache Mapping. */
        public function createRapidDiskCacheMapping($JSONRequest) {
                $nameObject = json_decode($JSONRequest, true);
                $rdsk = $nameObject['rapidDisk'];
                $source = $nameObject['sourceDrive'];
                $command = implode(' ', array($this->RDISK_CMD, '--cache-map', $rdsk, $source, "8"));
                exec($command, $output, $retval);
                $rtn = array('errorCode' => $retval, 'message' => implode(' ', $output));
                http_response_code(200); /* OK */
                print json_encode($rtn);
        }

        /* @DELETE: Remove an existing RapidDisk-Cache Mapping. */
        public function removeRapidDiskCacheMapping($disk) {
		$command = implode(' ', array($this->RDISK_CMD, '--cache-unmap', $disk));
		exec($command, $output, $retval);
                $rtn = array('errorCode' => $retval, 'message' => implode(' ', $output));
                http_response_code(200); /* OK */
                print json_encode($rtn);
        }
}
?>
